### PR TITLE
[RLlib] Support native tf.keras.Model (milestone toward obsoleting ModelV2 class).

### PR DESCRIPTION
**Pull request by sven1977**
This PR adds support for native keras models in TF- and eager Policies.

This is an experimental API extension that's fully backward compatible (ModelV2s will still work, of course!).

- Users can define a pure tf.keras.Model as their model and have the TF/eager policy use it via the "custom_model" config key.
- The keras model only has to implement a `get_initial_state` method (this requirement may be removed as well to support functional model building or using tf.keras.Sequential). All other ModelV2 methods are no longer required.
- The keras.Model `call` method must be implemented for the forward logic.
- An example (RNN) model class has been added (examples/models/modelv3.py), as well as a test case confirming that PPO is learning CartPole using this approach.
- A new (experimental) config flag "_use_default_native_models" (False by default) has been added to catalog.py. If True, RLlib will use built-in keras default models where possible (e.g. done in the PPO test). Setting this flag to True is save (RLlib will not try to use keras models where this is not supported, e.g. for torch, for VisionNets, etc..).

NOTE: This PR is for TF only. A solution for supporting pure native torch.Modules is to follow.

Follow up PRs will include:
- Making all default tf nets native keras (VisionNet + RNN + LSTM-/attention-wrappers).
- Support for PyTorch native Modules.
- Examples for custom Sequential types, RNNs (how to specify initial states?), trajectory view API, etc..

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/marwil/marwil_tf_policy.py [hunk 1]
```diff
@@ -67,7 +67,8 @@ def postprocess_advantages(policy,
         # input_dict.
         # Create an input dict according to the Model's requirements.
         index = "last" if SampleBatch.NEXT_OBS in sample_batch.data else -1
-        input_dict = policy.model.get_input_dict(sample_batch, index=index)
+        input_dict = sample_batch.get_single_step_input_dict(
```
Comment (sven1977) [inline rllib/agents/marwil/marwil_tf_policy.py:70]: Moved this helper method to SampleBatch (from ModelV2), where it's more fitting.

File: rllib/agents/ppo/ppo_tf_policy.py [hunk 2]
```diff
@@ -29,21 +29,29 @@
...
+            the loss for.
         dist_class (Type[ActionDistribution]: The action distr. class.
         train_batch (SampleBatch): The training data.
 
     Returns:
         Union[TensorType, List[TensorType]]: A single loss tensor or a list
             of loss tensors.
     """
-    logits, state = model.from_batch(train_batch)
+    if isinstance(model, tf.keras.Model):
```
Comment (sven1977) [inline rllib/agents/ppo/ppo_tf_policy.py:48]: New model API: native + return 3 values (model-out, state-outs, extra-outs (e.g. vf)).

File: rllib/agents/ppo/ppo_tf_policy.py [hunk 3]
```diff
@@ -86,7 +94,6 @@ def reduce_mean_valid(t):
 
     if policy.config["use_gae"]:
         prev_value_fn_out = train_batch[SampleBatch.VF_PREDS]
-        value_fn_out = model.value_function()
```
Comment (sven1977) [inline rllib/agents/ppo/ppo_tf_policy.py:89]: One reason for the new API is that we can get rid of these async `value_function` calls.

File: rllib/agents/ppo/tests/test_ppo.py [hunk 4]
```diff
@@ -82,6 +82,9 @@ def test_ppo_compilation_and_lr_schedule(self):
         # Settings in case we use an LSTM.
         config["model"]["lstm_cell_size"] = 10
         config["model"]["max_seq_len"] = 20
+        # Use default-native keras model whenever possible.
```
Comment (sven1977) [inline rllib/agents/ppo/tests/test_ppo.py:85]: Testing, whether this flag works (try using native keras default model).
Currently only works for FCNets.

File: rllib/examples/models/modelv3.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            initial_state=[
+                sample_batch["state_in_0"], sample_batch["state_in_1"]
+            ],
+        )
+        lstm_out = tf.reshape(lstm_out, [-1, lstm_out.shape.as_list()[2]])
+        logits = self.logits(lstm_out)
+        values = tf.reshape(self.values(lstm_out), [-1])
+        return logits, [h, c], {SampleBatch.VF_PREDS: values}
+
+    def get_initial_state(self):
```
Comment (sven1977) [inline rllib/examples/models/modelv3.py:53]: We should be able to get rid of this as well in the future via the trajectory view API.

File: rllib/models/modelv2.py [hunk 6]
```diff
@@ -333,71 +333,6 @@ def is_time_major(self) -> bool:
         """
         return self.time_major is True
 
-    # TODO: (sven) Experimental method.
```
Comment (sven1977) [inline rllib/models/modelv2.py:336]: Was moved to SampleBatch instead (makes more sense there).
SampleBatch converts itself into an single-timestep input-dict (which is also a SampleBatch).

File: rllib/policy/sample_batch.py [hunk 7]
```diff
@@ -588,6 +579,71 @@ def data(self):
                 old="SampleBatch.data[..]", new="SampleBatch[..]", error=False)
         return self
 
+    # TODO: (sven) Experimental method.
```
Comment (sven1977) [inline rllib/policy/sample_batch.py:582]: Moved here from ModelV2 (makes more sense inside SampleBatch).

File: rllib/agents/ppo/ppo_tf_policy.py (hunk 2 above)
Comment (michaelzhiluo) [reply rllib/agents/ppo/ppo_tf_policy.py:48]: Assuming ModelV2 will be deprecated soon :)

File: rllib/examples/models/modelv3.py (hunk 5 above)
Comment (michaelzhiluo) [reply rllib/examples/models/modelv3.py:53]: Some users would want different types of initial states (e.g. xavier initalization) right?

File: rllib/models/tf/fcnet.py [hunk 8]
```diff
@@ -127,3 +129,129 @@ def forward(self, input_dict: Dict[str, TensorType],
...
+        inputs = tf.keras.layers.Input(
+            shape=(int(np.product(input_space.shape)), ), name="observations")
+        # Last hidden layer output (before logits outputs).
+        last_layer = inputs
+        # The action distribution outputs.
+        logits_out = None
+        i = 1
+
+        # Create layers 0 to second-last.
+        for size in hiddens[:-1]:
```
Comment (michaelzhiluo) [inline rllib/models/tf/fcnet.py:181]: Is this code copy pasted from the prior modelv2 fcnet for tf?

**Review by michaelzhiluo — APPROVED** [review]

**Comment by smorad** [comment]
@sven1977 
This breaks `rllib` if `tensorflow` is not installed.

https://github.com/ray-project/ray/blob/e973b726c2e4039bd5a35ecfffad5028a15cdfee/rllib/models/catalog.py#L741
Should be if `if tf and issubclass(model_class, tf.keras.Model):`

and https://github.com/ray-project/ray/blob/e973b726c2e4039bd5a35ecfffad5028a15cdfee/rllib/models/catalog.py#L408
Should be `if tf and framework not in ["tf", "tf2", "tfe"]`

I'd patch this but I've managed to totally mangle my fork/local repo atm.

**Pull request merged**